### PR TITLE
[BEAM-38] Allow DoFns to be Reused

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Modify the documentation of ParDo to allow DoFns to be reused for
multiple bundles in the general case. This allows DoFns which implement
start and finish bundle to be reused.

Update ParDoTest.TestDoFn to succeed if reused.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles @dhalperi @davorbonaci @bjchambers

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 1]
```diff
@@ -121,7 +121,8 @@ public TestDoFn(List<PCollectionView<Integer>> sideInputViews,
 
     @Override
     public void startBundle(Context c) {
-      assertEquals(State.UNSTARTED, state);
+      // The Fn can be reused, but only if FinishBundle has been called.
+      assertThat(state, anyOf(equalTo(State.UNSTARTED), equalTo(State.FINISHED)));
```
Comment (davorbonaci) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: sounds like there should be more changes to actually try this out

**Comment by davorbonaci** [comment]
LGTM

**Comment by kennknowles** [comment]
Whoops, I didn't notice that an email discussion had been introduced. I just merged this since it was a known issue for so long and blocks progress on replacing the direct runner. I'm perfectly happy for things to change again if the dev list discussion reaches a new conclusion for [BEAM-38](https://issues.apache.org/jira/browse/BEAM-38).

**Pull request merged**